compiler/optimizer/VPConstraint.cpp

```diff
@@ -5797,6 +5904,24 @@ void TR::VPClass::print(TR::Compilation *comp, TR::FILE *outFile)
       return;
    if (_type)
       _type->print(comp, outFile);
```

Comment:
Since `VPClass._type._typeHintClass` might not be the same as `VPClass._typeHintClass`, wouldn't it be useful to print out both to help debug intersect/merge related issues?

Reply:
Once a `VPClassType` constraint is bundled into a larger `VPClass` constraint, the hint of the `VPClassType` should be irrelevant. But I suppose that you're right that it's possible for it to accidentally be relevant in a way that affects a bug. And I suppose also that some consuming code could (though it shouldn't) look for a hint using (effectively) `c->getClassType()->getTypeHintClass()` instead of `c->getTypeHintClass()`. And all this does suggest that we'll want to see both hints at least sometimes... :thinking: How about this?
- Distinguish the two hints in some way other than ordering, e.g. `(*hint ...)` for `VPClass` and `(hint ...)` for `VPClassType`. When only one is shown, this tells us which.
- Print the `VPClass` hint whenever (and only when) it's different from the one implied by `_type` (which is *top* when there is no `_type`). Print it every time it differs, even if the `VPClass` hint is *top*.

These rules would give us the following cases:
- No hint printed → no hint in the constraint, i.e. all *top*. I think this should be the most common.
- `(hint ...)` only → `VPClass` and `VPClassType` both agree on the hint.
- `(*hint ...)` only → `VPClass` has a hint but `VPClassType` has *top*/no hint.
- `(hint ...) (*hint ...)` → `VPClassType` and `VPClass` disagree. I think this should be the least common.

When the last case occurs, I expect it would usually look like `(hint 0x...) (*hint <bottom>)`.

What do you think?

Reply:
Good idea! If I understand it correctly, `TR::VPClass::print` would look like something as below. I changed `"*"` to `"+"` to avoid it from meaning we are dereferencing a pointer

```c++
=== TR::VPClass::print ===

   if (_typeHintClass)
      {
      TR_OpaqueClassBlock *typeHintClassFromVPClassType = _type ? _type->getTypeHintClass() : NULL;
      if (_typeHintClass != typeHintClassFromVPClassType) 
         {
         trfprintf(outFile, " (+hint 0x%p", _typeHintClass);
         if (!isSpecialClass((uintptr_t)_typeHintClass))
            {
            int32_t len;
            const char *sig = TR::Compiler->cls.classSignature_DEPRECATED(comp, _typeHintClass, len, comp->trMemory());
            trfprintf(outFile, " %.*s)", len, sig);
            }
         else
            {
            trfprintf(outFile, " <bottom>)");
            }
         } 
      }
```

Reply:
Yes, that looks good :+1:

Thanks!

Reply:
Addressed in bb6ab93